services/api-server/tests/unit/test_services_rabbitmq.py

```diff
@@ -0,0 +1,327 @@
+    await asyncio.sleep(1)
+    assert consumer_message_handler.await_count == 1
+    (data,) = consumer_message_handler.call_args[0]
+    assert isinstance(data, bytes)
+    received_message = LoggerRabbitMessage.parse_raw(data)
+
+    assert received_message.user_id == user_id
+    assert received_message.project_id == project_id
+    assert received_message.node_id == node_id
+    assert received_message.messages == ["expected message"] * 3
+
+
+#
+# --------------------
+#
+
+
+class JobLog(BaseModel):
+    job_id: ProjectID
+    node_id: NodeID | None
+    log_level: LogLevelInt
+    messages: list[LogMessageStr]
+
+
+@pytest.fixture
+def new_routes_injected(app: FastAPI):
+    # https://docs.python.org/3/library/asyncio-queue.html#queue
+
+    _NEW_LINE = "\n"
+
+    async def get_buffer(request: Request):
+        return asyncio.Queue()
+
+    async def listen_to_logs(
+        buffer: Annotated[asyncio.Queue, Depends(get_buffer)],
+        project_id: ProjectID,
+        follow: bool = False,
+    ):
+        async def _consume_log(data: bytes):
+            got = LoggerRabbitMessage.parse_raw(data)
+            assert got.project_id == project_id
+
+            item = JobLog(
+                job_id=got.project_id,
+                node_id=got.node_id,
+                log_level=got.log_level,
+                messages=got.messages,
+            )
+
+            if follow:
+                await buffer.put(item)
+            else:
+                with suppress(asyncio.QueueFull):
+                    buffer.put_nowait(item)
+
+            return True
+
+        # listen to project_id
+        async with rabbit_consuming_context(app, project_id, _consume_log):
+
+            yield
+
+            print("stopping listening to logs")
+
+    async def _json_logs_generator(
+        queue: asyncio.Queue, follow: bool
+    ) -> AsyncIterable[str]:
+        if follow:
+            # forever
+            while True:
+                # if queue is empty, it waits
+                job_log = await queue.get()
+                yield job_log.json() + _NEW_LINE
+        else:
+            try:
+                await asyncio.sleep(0)
+                # Return an item if one is immediately available
+                job_log = queue.get_nowait()
+                yield job_log.json() + _NEW_LINE
+
+            except asyncio.QueueEmpty:
+                pass
+
+    @app.get("/projects/{project_id}/logs")
+    async def _stream_logs_handler(project_id: ProjectID, *, follow: bool = False):
+        buffer = asyncio.Queue()
+
+        async def _consume_log(data: bytes):
+            got = LoggerRabbitMessage.parse_raw(data)
+            assert got.project_id == project_id
+
+            item = JobLog(
+                job_id=got.project_id,
+                node_id=got.node_id,
+                log_level=got.log_level,
+                messages=got.messages,
+            )
+
+            if follow:
+                await buffer.put(item)
```

Comment:
Hi Pedro,
Thanks a lot for the explanation of all of this today. I think it looks very good already. I had a look at what I could find about this problem online and I think it is likely we can avoid the buffer (at least if I understand the sources I found correctly). I think this is the first approach I will try to implement. From https://datatracker.ietf.org/doc/html/rfc6202#section-3 it seems to me that the client will not make requests to the server to get subsequent "chunks". So if I understand that correctly `_consume_logs` does not need to store the `JobLog`, but can send it directly to the client as a `StreamingResponse`. Then of course the client would have to do some buffering. But I would hope this is implemented in the `httpx` client. If you already played with this approach(?) it of course doesn't make sense for me to try that, but if not then I will try that out.